### PR TITLE
Use abstract classes to avoid repeated code

### PR DESCRIPTION
**Pull request by SaralynOgden**
Moves the code that is used in both mt5 and simulator classes to abstract classes (derived from the interfaces) to avoid repeating code.

### CODE REVIEW

**Review by bpsix — APPROVED** [review]
Looks good to me. What are your thoughts on cleaning up our file structure a bit? We could place related logic in subdirectories, or combine related classes into a more generally named .py file.

**Review by BensGitHub2022 — CHANGES_REQUESTED** [review]
The stated design goal is contract by interface and dependency injection. This change set removes almost all interfaces. How would this work?

Here is a link to design doc that I drafted: https://docs.google.com/document/d/1iiVMiqNZ1Z_IIrFy1lsWsNMpTv58TySWihE-_jt4TwA/edit

- I think some things named slightly differently when I wrote this but the gist is symbol_feed / ticker_symbol = "symbol"
- if you scroll down to "In the composition root" section there is example of how interfaces are called

**Comment by SaralynOgden** [comment]
> The stated design goal is contract by interface and dependency injection. This change set removes almost all interfaces. How would this work?
> 
> Here is a link to design doc that I drafted: https://docs.google.com/document/d/1iiVMiqNZ1Z_IIrFy1lsWsNMpTv58TySWihE-_jt4TwA/edit
> 
> * I think some things named slightly differently when I wrote this but the gist is symbol_feed / ticker_symbol = "symbol"
> * if you scroll down to "In the composition root" section there is example of how interfaces are called

@BensGitHub2022 Ok. Yes, we still need the interfaces in order to implement dependency injection. We still will want to avoid code reuse to keep our code DRY. I see two options for using dependency injection and not having the code repeat:
1. We can make all of the abstract classes inherit from the interface which would return us to a state where all of the classes are ultimately inheriting from the interfaces
2. We can abstract the classes out further to have other classes which hold on to the the common code that sits in mock and mt5 and use composition to get the common methods in both.

Since an immediate answer as to what the unique purpose of the new classes would be in 2, I think 1 is the more logical solution.

**Comment by BensGitHub2022** [comment]
Where I am coming from this is why do we need to remove interfaces in the first place or reintroduce them to accommodate abstract class?

As far as the case for abstract classes, I am not following the notion of saving on typing / reuse of code has to do with this. It would assume that in addition to shared behavior some of these classes also have shared data. For example, the account class - what commonality does the sim_account class share with the real account class that is not already expressed through the interface (i.e. its behavior)?

File: src/abstract_trade_execution.py [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+    
+    RISK = .02
+    current_risk_per_trade = 0.0
+    current_lot_size = 0.0
+    account_info: object
+    
+    def __init__(self, account: object):
+        self.account_info = account
+        
+    def calc_risk_per_trade(self) -> float:
```
Comment (SaralynOgden) [inline src/abstract_trade_execution.py:13]: *

File: src/abstract_trade_execution.py [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+        self.current_risk_per_trade = self.account_info.get_account_balance() * self.RISK
+        return self.current_risk_per_trade
+
+    def calc_lot_size(self, price: float) -> float:
```
Comment (SaralynOgden) [inline src/abstract_trade_execution.py:17]: *

File: src/abstract_trade_execution.py [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+        self.current_risk_per_trade = self.calc_risk_per_trade()
+        self.current_lot_size = self.current_risk_per_trade / price
+        return self.current_lot_size
+
+    def close_all_positions(self, bid, ask, deviation) -> bool:
```
Comment (SaralynOgden) [inline src/abstract_trade_execution.py:22]: *

File: src/abstract_trade_execution.py [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+        positions = self.account_info.get_positions()
+        for position in positions:
+            self.close_position(position, bid, ask, deviation)
+    
+    def do_nothing(self):
```
Comment (SaralynOgden) [inline src/abstract_trade_execution.py:27]: *

**Comment by SaralynOgden** [comment]
> Where I am coming from this is why do we need to remove interfaces in the first place or reintroduce them to accommodate abstract class?
> 
> As far as the case for abstract classes, I am not following the notion of saving on typing / reuse of code has to do with this. It would assume that in addition to shared behavior some of these classes also have shared data. For example, the account class - what commonality does the sim_account class share with the real account class that is not already expressed through the interface (i.e. its behavior)?

I was looking to define the common fields and methods between the classes in the abstract class.

Defining the fields provides a definition of expected, common fields that should be used in each class. While I prefer the style to give hints as how to build out a new class, it is not terribly important. This means that 3/4 of the abstract classes could be disregarded.

Defining the common methods in the abstract class is more important since, we want these methods to behave the *exact* same way if the data is MT5, mocked, or any other data source we bring in. Further, if we'd like to change the implementation of the common methods, it is much easier and safer to do so when they are located in the same abstract class (or class that composes them). The classes that repeat methods are the trade execution classes. We should define these methods in one place either through an abstract class, some sort of helper, or by putting them in a different class altogether (maybe the trade bot). I have commented dots next to the methods that are common between the two trade execution classes in this PR.

**Comment by BensGitHub2022** [comment]
Common behavior is expressed and captured by the interface. I don't understand how abstract class would improve over this. In my original question, what data members are common between sim account and real account or for that matter the trade executor class?

**Comment by SaralynOgden** [comment]
> Common behavior is expressed and captured by the interface. I don't understand how abstract class would improve over this. In my original question, what data members are common between sim account and real account or for that matter the trade executor class?

Data does not have to be the same between two classes for them to need a common base class. If the methods are implemented the same between the classes and the classes serve the same functionality overall, then they should have a common base class that defines all of the methods that are the same.

For example, [this code in the mt5 class](https://github.com/BensGitHub2022/MT5-Expert-Advisor/blob/main/src/trade_execution_mt5.py#L18) is exactly the same as [this code in the sim class](https://github.com/BensGitHub2022/MT5-Expert-Advisor/blob/main/src/trade_execution_sim.py#L24) and the code would also be the same in a class for a coinbase class. So, therefore, it should be defined in one place. An abstract class fulfills that need.

**Comment by BensGitHub2022** [comment]
This comment can be interpreted several ways (I see what you are saying that derived abstract classes can have different data fields but their base class in general pass the same data between the two). The original part of my question was driving at what data fields are shared between simulated account fields and mt5 account fields (assuming they were kept in the same base class)? 

I think for your example, you have picked the one function that incidentally is shared between two classes. The best way to deal with this is factor it out into a helper function which I have done. See the updated PR, I don't think finding some shared code between a few classes will merit switching to abstract classes

**Comment by SaralynOgden** [comment]
There are four methods shared between trade execution classes. I have put comments next to each in this PR.

It is best practice to limit repeated code as much as possible.

**Comment by BensGitHub2022** [comment]
we can factor all of those methods out, I will update my PR

**Comment by BensGitHub2022** [comment]
I refactored calc_risk_per_trade and calc_lot_size to shared_helper_functions.py -> consider making this into some kind of "calculator" class because I think there is potential later functionality here.

the close_all_positions and do_nothing have very basic functionality so I left them alone -> close_all_positions basically just iterates a collection, do_nothing does nothing currently but may eventually report different items in sim class vs live implementation so I left it out as well

**Pull request closed without merging**